Add new nextcloud chart

affix opened the pull request:
<!--
Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

**What this PR does / why we need it**:
This PR adds a nextcloud chart to incubator 

Nextcloud offers industry-leading on-premises file sync and online collaboration technology. Our expertise is in combining the convenience and ease of use of consumer-grade solutions like Dropbox and Google Drive with the security, privacy and control business needs.

**Special notes for your reviewer**:
This is loosely based around the owncloud chart, However I as many other prefer to use nextcloud for some of its optimisations.

Uses the official Nextcloud Docker image

Commit:
Add new nextcloud chart

affix commented:
/assign @mattfarina

bbriggs commented:
I was literally writing a chart for that this weekend. Glad to see someone else took an interest too.

unguiculus commented:
/assign

mattfarina commented on incubator/nextcloud/Chart.yaml:
```diff
@@ -0,0 +1,19 @@
+sources:
+- https://github.com/nextcloud/server
+maintainers:
+- name: Keiran Smith
```
The name should be your github username. That helps us to know who to contact when there's an issue with the chart.

mattfarina commented on incubator/nextcloud/OWNERS:
```diff
@@ -0,0 +1,12 @@
+- tompizmor
+- sameersbn
+- carrodher
+- affix
```
For the time you can drop this file. It looks like you copied this from a chart managed by Bitnami. Since you're not a member of the Kubernetes org at the moment this isn't going to enable you to review or merge for the chart. Once you have some more contributions you can apply for membership and then a file like this will make sense.

mattfarina commented on incubator/nextcloud/values.yaml:
```diff
@@ -0,0 +1,184 @@
+resources:
+  requests:
+    memory: 512Mi
+    cpu: 300m
```
> It is generally not advisable to provide hard resource limits to workloads and leave it configurable unless the workload requires such quantity bare minimum to function.

This is from the r[eviewers guide](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#kubernetes-native-workloads).

Can you comment these out unless this is the minimum to function.

mattfarina commented on incubator/nextcloud/values.yaml:
```diff
@@ -0,0 +1,184 @@
+##
+ingress:
+  enabled: false
+  servicePort: 80
```
Does the container image allow for both http and https? It might be nice to allow for secure out of the box. That doesn't need to be part of the first PR but would be nice to add.

mattfarina commented on incubator/nextcloud/values.yaml:
```diff
@@ -0,0 +1,184 @@
+image:
+  registry: docker.io
+  repository: nextcloud
+  tag: latest
```
Can you update the image information to follow our standard layout? See [the guide](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#configuration) for more information and an example.

Also, please use a tagged version rather than latest. This will enable consistency between environments.

mattfarina commented on incubator/nextcloud/templates/ingress.yaml:
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.ingress.enabled }}
+apiVersion: {{ required "A valid .Values.networkPolicyApiVersion entry required!" .Values.networkPolicyApiVersion }}
```
Why is the API version a values parameter?

mattfarina commented on incubator/nextcloud/templates/deployment.yaml:
```diff
@@ -0,0 +1,103 @@
+{{- if include "nextcloud.host" . -}}
+apiVersion: extensions/v1beta1
```
Can you use the `apps/v1beta2` version instead?

mattfarina commented on incubator/nextcloud/requirements.yaml:
```diff
@@ -0,0 +1,5 @@
+dependencies:
+- name: mariadb
+  version: 0.7.0
```
The version supports ranges. What about `~0.7.0` so that it can pull in patch level differences?

mattfarina reviewed (commented):
Thanks for submitting the pull request. I know of several people who want to use nextcloud on Kubernetes. This is great.

I did a quick review to provide some feedback.

unguiculus commented:
Please compare with what you get from `helm create` using Helm 2.8+ and update accordingly.

pierreozoux commented:
I think it would make more sense to PR against `stable` folder. It is a pretty straight forward addition.

Also I can help maintain this chart.

Another question: how could we keep this under `/nextcloud` github namespace, and sync it with the official charts? This would help a lot the pace of development.

affix committed:
Move nextcloud to stable

affix committed:
Changes from Code Review

- Change maintainer to affix
- Remove OWNERS file
- Comment out resources in values.yml
- Default to nextcloud 13.0.4-apache
- Specify Ingress api version
- Make Deployment v1beta2
- Support patch level MariaDB

Signed-off-by: Keiran Smith <contact@keiran.scot>

affix committed:
Changes from Code Review

- Change maintainer to affix
- Remove OWNERS file
- Comment out resources in values.yml
- Default to nextcloud 13.0.4-apache
- Specify Ingress api version
- Make Deployment v1beta2
- Support patch level MariaDB

Signed-off-by: Keiran Smith <contact@keiran.scot>

affix committed:
Merge branch 'nextcloud' of github.com:affix/charts into nextcloud

affix committed:
remove network policy values

Signed-off-by: Keiran Smith <contact@keiran.scot>

affix committed:
Minor Fixes to Deployment + Service

unguiculus commented on stable/nextcloud/README.md:
```diff
@@ -0,0 +1,126 @@
+
+## Prerequisites
+
+- Kubernetes 1.4+ with Beta APIs enabled
```
Should be 1.9+

unguiculus commented on stable/nextcloud/templates/deployment.yaml:
```diff
@@ -0,0 +1,103 @@
+{{- if include "nextcloud.host" . -}}
+apiVersion: extensions/v1beta1
```
apps/v1

unguiculus commented on stable/nextcloud/.helmignore:
```diff
@@ -0,0 +1 @@
+.git
```
Use `.helmignore` you get from `helm create`

unguiculus commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,180 @@
+    # - secretName: nextcloud.cluster.local
+    #   hosts:
+    #     - nextcloud.cluster.local
+
```
I'd suggest you add one level of nesting:

```yaml
nextcloud:
  host:
  loadBalancerIP:
  username:
  password:
  ...
```

unguiculus reviewed (changes requested):
Again, please compare with what you get from helm create using Helm 2.8+ and update accordingly (`_helpers.tpl`, standard labels, etc.).

Also, you'll need to merge in the latest master to make CI happy.

billimek committed:
changed referenced in original pull request

billimek committed:
fixing filename

affix committed:
Update MariaDB

affix committed:
Merge pull request #1 from billimek/nextcloud_updates

changed referenced in original pull request

affix committed:
Merge branch 'master' of github.com:helm/charts into nextcloud

affix committed:
Update Helmignore

affix committed:
Use apps/v1

affix committed:
PR Changes

tilosp commented on stable/nextcloud/README.md:
```diff
@@ -0,0 +1,128 @@
+# nextcloud
+
+[nextcloud](https://nextcloud.org/) is a file sharing server that puts the control and security of your own data back into your hands.
```
this should be https://nextcloud.com/

mattfarina replied:
Good catch. https://nextcloud.org gives a certificate error.

tujian commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,170 @@
+  email: user@example.com
+
+## Set to `yes` to allow the container to be started with blank passwords
+## ref: https://github.com/bitnami/bitnami-docker-nextcloud#environment-variables
```
This ref [https://github.com/bitnami/bitnami-docker-nextcloud#environment-variables](https://github.com/bitnami/bitnami-docker-nextcloud#environment-variables), which is substituted from [https://github.com/bitnami/bitnami-docker-owncloud#environment-variables](https://github.com/bitnami/bitnami-docker-owncloud#environment-variables), leads to 404.

anarcat replied:
```suggestion
## ref: https://github.com/nextcloud/docker#auto-configuration-via-environment-variables
```

giacope commented on stable/nextcloud/Chart.yaml:
```diff
@@ -0,0 +1,19 @@
+name: nextcloud
+version: 0.1.1
```
Why is the chart version `0.1.1`? Shouldn't it be  `0.1.0` since it's a new chart?

giacope commented on stable/nextcloud/Chart.yaml:
```diff
@@ -0,0 +1,19 @@
+name: nextcloud
+version: 0.1.1
```
So apparently, charts going to stable need to start at `1.0.0`

desaintmartin replied:
```suggestion
version: 1.0.0
```

anarcat commented on stable/nextcloud/README.md:
```diff
@@ -0,0 +1,128 @@
+# nextcloud
+
+[nextcloud](https://nextcloud.org/) is a file sharing server that puts the control and security of your own data back into your hands.
```
making this formal with GH's new proposed change gizmo:

```suggestion
[nextcloud](https://nextcloud.com/) is a file sharing server that puts the control and security of your own data back into your hands.
```

anarcat commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,170 @@
+## Bitnami nextcloud image version
+## ref: https://hub.docker.com/r/bitnami/nextcloud/tags/
```
```suggestion
## Official nextcloud image version 
## ref: https://hub.docker.com/r/library/nextcloud/tags/
```

anarcat commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,170 @@
+  enabled: true
+
+  ## Create a database
+  ## ref: https://github.com/bitnami/bitnami-docker-mariadb/blob/master/README.md#creating-a-database-on-first-run
```
another broken link:

```suggestion
  ## ref: https://github.com/nextcloud/docker#using-an-external-database
```

anarcat commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,170 @@
+  mariadbDatabase: nextcloud
+
+  ## Create a database user
+  ## ref: https://github.com/bitnami/bitnami-docker-mariadb/blob/master/README.md#creating-a-database-user-on-first-run
```
just remove this duplicate link

```suggestion
```

anarcat commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,170 @@
+  mariadbUser: nextcloud
+
+  ## Password for mariadbUser
+  ## ref: https://github.com/bitnami/bitnami-docker-mariadb/blob/master/README.md#creating-a-database-user-on-first-run
```
just remove this duplicate link

```suggestion
```

anarcat commented on stable/nextcloud/values.yaml:
```diff
@@ -0,0 +1,170 @@
+  # mariadbPassword:
+
+  ## MariaDB admin password
+  ## ref: https://github.com/bitnami/bitnami-docker-mariadb/blob/master/README.md#setting-the-root-password-on-first-run
```
just remove this duplicate link

```suggestion
```

anarcat reviewed (changes requested):
so the bitnami docker image is gone, as others have noted here. i've tried to fix all the 404s refering to that image and point instead at the official docker image managed by the nextcloud organization. i am not sure it's sufficient here because the newer image might not work the same way esp. WRT the mariadb bitnami image.
